### ISSUE TITLE
consul-replicate not replicating all keys

### ISSUE THREAD

**Issue by nimbusscale** [hashicorp/consul-replicate#29]
We are looking at possible using consul-replicate to replicate data between data centers, but in testing we have found that not all keys are replicated, even though consul-replicate seems to think they are. Hopefully it's something with our setup, but maybe there is some other issue?

For testing we have a python script that writes as many keys as it can to a specified prefix within a specified number of seconds (i.e. write as many keys to sc-poc/ prefix as you can for 10 seconds). This script can be seen here:  [consul_benchmark.py](https://gist.github.com/nimbusscale/67dc8ce54786fc3c847a)

In the latest run this script was able to write 769 keys in 10 seconds, but only 457 of those keys were replicated. 

The following commanded was used for the replication (debug added once we realized keys were missing):

    consul-replicate -prefix "sc-poc/@us-east-1" -log-level debug > /tmp/repl.log 2>&1 &

consul-replicate was started before any keys in the prefix existed and was left running a while (10 - 15 min) after the writing of keys had stopped, so it wasn't like it was delayed.

Here is the output of an ipython session used to try and figure out what happened during the test.

The following shows me connecting to both the local cluster (source) and remote cluster (dest) and getting a count of keys written to the "sc-poc"/ prefix.

	In [1]: import consul
	In [2]: lc = consul.Consul(host='127.0.0.1')
	In [3]: rc = consul.Consul(host='10.0.0.141')
	In [4]: lkeys = lc.kv.get("sc-poc/",recurse=True)
	In [5]: len(lkeys[1])
	Out[5]: 769
	In [6]: rkeys = rc.kv.get("sc-poc/",recurse=True)
	In [7]: len(rkeys[1])
	Out[7]: 457

Now I get a list of keys from both local and remote and generate a list of keys missing on the remote side. 312 keys missing (312 + 457 = 769)

	In [8]: lkeylist = []
	In [9]: for key in lkeys[1]:
	   ...:         lkeylist.append(key['Key'])
	   ...:
	In [10]: rkeylist = []
	In [11]: for key in rkeys[1]:
	   ....:        rkeylist.append(key['Key'])
	   ....:
	In [12]: missing_keys = list(set(lkeylist) - set(rkeylist))
	In [13]: len(missing_keys)
	Out[13]: 312

Now I get a single missing keys and view the key both local and remote to validate the key is indeed missing on the remote. 

	In [14]: sample_key = missing_keys[0]
	In [15]: sample_key
	Out[15]: u'sc-poc/ip-10-1-0-5/key_238'
	In [16]: lc.kv.get(sample_key)
	Out[16]:
	('72603',
	 {u'CreateIndex': 72603,
	  u'Flags': 0,
	  u'Key': u'sc-poc/ip-10-1-0-5/key_238',
	  u'LockIndex': 0,
	  u'ModifyIndex': 72603,
	  u'Value': '74107ca52f9ee603783d8d40613bc399eae203bd'})
	In [17]: rc.kv.get(sample_key)
	Out[17]: ('43347', None)

Next I look at the "service/consul-replicate/statuses" key to determine what key consul-replicate thinks is the last key it replicated. I then look for that key on both the local and remote.

	In [18]: rc.kv.get("service",recurse=True)
	Out[18]:
	('43347',
	 [{u'CreateIndex': 42850,
	   u'Flags': 0,
	   u'Key': u'service/consul-replicate/statuses/ba7a6cd8f91c8bac2bdcff42cd08df81',
	   u'LockIndex': 0,
	   u'ModifyIndex': 43347,
	   u'Value': '{\n  "LastReplicated": 73133,\n  "Source": "sc-poc/",\n  "Destination": "sc-poc/"\n}'}])
	In [19]: for key in lkeys[1]:
	   ....:        if key['CreateIndex'] == 73133:
	   ....:                        print(key)
	   ....:
	{u'LockIndex': 0, u'ModifyIndex': 73133, u'Value': '09565407012308833ea63643c958f185b20dd3a7', u'Flags': 0, u'Key': u'sc-poc/ip-10-1-0-5/key_768', u'CreateIndex': 73133}
	In [20]: rc.kv.get('sc-poc/ip-10-1-0-5/key_768')
	Out[20]: ('43347', None)

Lastly looking the debug output (attached [repl.log.txt](https://github.com/hashicorp/consul-replicate/files/39402/repl.log.txt)) for "sc-poc/ip-10-1-0-5/key_238" and "sc-poc/ip-10-1-0-5/key_768" it states that consul-replicate thinks those keys have indeed been replicated.

	[ec2-user@ip-10-0-0-141 ~]$ grep sc-poc/ip-10-1-0-5/key_238 /tmp/repl.log
	2015/11/19 18:27:05 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:05 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:06 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:06 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:06 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:07 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:07 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:08 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:08 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:08 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:09 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:09 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:10 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:10 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:10 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:10 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:11 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:11 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:12 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	2015/11/19 18:27:12 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_238" is already replicated
	[ec2-user@ip-10-0-0-141 ~]$ grep sc-poc/ip-10-1-0-5/key_768 /tmp/repl.log
	2015/11/19 18:27:12 [DEBUG] (runner) skipping because "sc-poc/ip-10-1-0-5/key_768" is already replicated
	[ec2-user@ip-10-0-0-141 ~]$

Given that these keys are not in the remote side, something fishy seems to be going on.

Keys seem to be missing every run of a test when we write a few hundred keys.

It doesn't look like this issue is seen if replication is started after the writing of keys has been completed. For example is we write a bunch of keys, and then start consul-replicate everything seems to work fine.

    $ consul-replicate -version
    consul-replicate v0.2.0

**Comment by nimbusscale** [comment]
I performed another test where I created a single key in the "sc-pc/" prefix prior to starting replication. This was intended to see if the prefix existing prior to replication starting made a difference.

There was no change and some keys were still missing on the remote side.

**Comment by evan2645** [comment]
i have seen similar behavior. it occurs only when the consul servers are under load. i have only found it today, so am still investigating, but they are the same symptoms

**Comment by slackpad** [comment]
I think I've found the problem here - it looks like there can be multiple `replicate()` goroutines running for the same prefix if changes are coming in rapidly, which could cause some hazards with the last replicated index. Trying to confirm this now and will work up a fix if that's indeed the issue.

@nimbusscale thanks so much for the repro scripts - those are super helpful.

**Comment by slackpad** [comment]
Ok I've got some PRs put together that fix this. Turns out it wasn't multiple `replicate()` calls interfering, but the view could update [`LastIndex`](https://github.com/hashicorp/consul-replicate/blob/master/runner.go#L349) in parallel, causing us to permanently skip entries.

**Comment by ssenaria** [comment]
I'm running into this issue w/ the codebase I just built from `master.`

    [root@indeploy001.atl consul-replicate]# ./consul-replicate -version
    consul-replicate v0.2.0-dev

It's skipping keys that do not exist in my destination DC. The interesting part is, the keys it skipped were deleted when I ran the newly compiled consul-replicate. Previously, I was using the binary from Feb. Things were working fine but I had an issue with the program not sending logs to syslog via the config file (see #46 ). I was asked to compile from `master` which fixed my issue but now this issue crops up.

**Comment by slackpad** [comment]
Hi @ssenaria is it possible you have multiple consul-replicate processing running at the same time? This is a super common way to get skipped keys.

**Comment by ssenaria** [comment]
Nope, but here's something interesting. If I run using the config file flag, it skips and/or deletes my keys. If I run and pass the flags via the command line it works fine.

Using `config`

    ./consul-replicate -config=/opt/consul-replicate/config.json

Using CLI flags works fine.

    ./consul-replicate -prefix="deploylist@phl" -consul "ocd.atl.ariasystems.net:8500" -syslog -log-level=debug

Also, when running the app using a config flag, the prefix for `apichecks` is coming through incorrectly. The keys should be `apichecks/worldpay/<KEY>`

    2016/06/02 14:42:07 [DEBUG] (runner) skipping because "apichecksworldpay/" is already replicated
    2016/06/02 14:42:07 [DEBUG] (runner) skipping because "apichecksworldpay/api_desc" is already replicated
    2016/06/02 14:42:07 [DEBUG] (runner) skipping because "apichecksworldpay/api_type" is already replicated


My config:

    consul = "127.0.0.1:8500"
    retry = "10s"
    log_level = "debug"
    max_stale = "10m"

    syslog {
        enabled = true
    }

    prefix {
        source = "apichecks@phl"
    }

    prefix {
        source = "deploylist@phl"
    }

**Comment by slackpad** [comment]
@ssenaria hmm - this might be an issue with an implicit / or something similar. Do you mind opening a new issue so we can work this over there, since I don't think this is recurrence of the race condition issue.
